### ISSUE TITLE
Two databases needed but only one in use?

### ISSUE THREAD

**Issue by piiskop** [ONLYOFFICE/DocumentServer#943]
**Do you want to request a *feature* or report a *bug*?**
Probably a bug.
**What is the current behavior?**
I [installed using DEB-file](https://medium.com/arvutiteadus/foss-document-editing-in-nextcloud-ae9fbf98bcf3). Why are there two databases needed, postgresql and (mariadb or mysql)? All of them are required dependencies for the DEB-file. Pavel Lobashov wrote that only postgresql was in use.
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
```
peacecop@peacecop-HP-ProBook-4520s:~/Allalaadimised$ sudo dpkg -i onlyoffice-documentserver_amd64.deb 
[sudo] password for peacecop: 
Selecting previously unselected package onlyoffice-documentserver.
(Reading database ... 343099 files and directories currently installed.)
Preparing to unpack onlyoffice-documentserver_amd64.deb ...
Unpacking onlyoffice-documentserver (5.6.0-17) ...
dpkg: dependency problems prevent configuration of onlyoffice-documentserver:
 onlyoffice-documentserver depends on mysql-client | mariadb-client; however:
  Package mysql-client is not installed.
  Package mariadb-client is not installed.
 onlyoffice-documentserver depends on nginx-extras (>= 1.3.13); however:
  Package nginx-extras is not installed.
 onlyoffice-documentserver depends on postgresql-client (>= 9.1); however:
  Package postgresql-client is not installed.
 onlyoffice-documentserver depends on pwgen; however:
  Package pwgen is not installed.
 onlyoffice-documentserver depends on supervisor (>= 3.0b2); however:
  Package supervisor is not installed.
 onlyoffice-documentserver depends on xvfb; however:
  Package xvfb is not installed.
dpkg: error processing package onlyoffice-documentserver (--install):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 onlyoffice-documentserver

```**What is the expected behavior?**
That neither _mariadb_ nor _mysql_ is listed as a dependency if they're not used.
**Did this work in previous versions of DocumentServer?**
No. Previously, DEB-file was prescribed to remove _mariadb_ completely in favor of _mysql_.
**DocumentServer version:**
ONLYOFFICE/5.6.0.17
**Operating System:**
_Ubuntu_

**Comment by piiskop** [comment]
I'm confused. Is ONLYOFFICE FOSS or not? Is it partly FOSS? Why are there private repos if its fully FOSS?

As we see, the source code needs more documentation.

**Comment by ShockwaveNN** [comment]
> Is ONLYOFFICE FOSS or not?

I don't think we ever stated that all version of our product fully Open Sourced

You can build ONLYOFFICE DocumentServer from source, but we got some private repos and I think we got that right

**Comment by ShockwaveNN** [comment]
After some discussion with dev I have a clear picture

`mysql-client | mariadb-client` is basically same thing, so we need one of them

In `posinst` script (you can look at it inside deb file) user can select which db to use, so we do not know db type on `dependency installation` step, only on `postint` step later

So if we remove hard support of `mysql-client`
User can install our package, select `mysql` in `postinst` and get a failure

So we require both

I'll close this issue since we do not planning remove both dependencies in near future.

**Comment by piiskop** [comment]
I don't understand your explanation.

I know that _mysql-client_ and _mariadb-clien_ are basically the same thing. The main difference is that the first one is a client to a proprietary software for what backwards compatibility isn't guaranteed and the the second one is a client for a FOSS for that backwards compatibilidy is guaranteed by its creator.

I also had a look inside DEB-file and knew that there was the requirement for at least one of them. You didn't answer my questions why is one of them needed if they're not used. Is it for backwards compatibility? If so then _mysql-client_ can't guarantee it anyway as I just wrote.

You tell me that the user can choose between _psql_ and (_mysql_ or _mariadb_). This is not true. _psql_ is forced. I just did the installation and I wasn't asked for anything. It would be nice of course if user will be asked for which database engine they want to use. However, this isn't happening.

I used _apt_ for installing. Is this the reason why _postgres_ as the default value was selected without asking me? Does _apt-get_ ask? I haven't found anywhere that there's that difference. So I wanted to test it out. First, I wanted to completely remove _onlyoffice-documentserver_:
`peacecop@peacecop-HP-ProBook-4520s:~$ sudo apt purge onlyoffice-documentserver 
[sudo] password for peacecop: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages will be REMOVED:
  onlyoffice-documentserver*
0 upgraded, 0 newly installed, 1 to remove and 5 not upgraded.
After this operation, 1 004 MB disk space will be freed.
Do you want to continue? [Y/n] 
(Reading database ... 366676 files and directories currently installed.)
Removing onlyoffice-documentserver (5.6.0-17) ...
Preparing for shutdown, it can take a lot of time, please wait...Done
Stopping documentserver services...
ds:spellchecker: stopped
ds:docservice: stopped
ds:converter: stopped
ds:metrics: stopped
Processing triggers for libc-bin (2.31-0ubuntu9) ...
(Reading database ... 354381 files and directories currently installed.)
Purging configuration files for onlyoffice-documentserver (5.6.0-17) ...
ERROR:  must be owner of schema public
dpkg: error processing package onlyoffice-documentserver (--purge):
 installed onlyoffice-documentserver package post-removal script subprocess returned error exit status 1
Errors were encountered while processing:
 onlyoffice-documentserver
E: Sub-process /usr/bin/dpkg returned an error code (1)
`
So it can't be completely removed? Maybe it just removed it but not completely:
`peacecop@peacecop-HP-ProBook-4520s:~$ sudo apt remove onlyoffice-documentserver 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package 'onlyoffice-documentserver' is not installed, so not removed
0 upgraded, 0 newly installed, 0 to remove and 5 not upgraded.
`
Oh yes, it did. But _Ubuntu_ complained about a system error and asked me whether I want to send an error report. I chose to do so. And what I saw was that:

> Crash report

> Sorry, a problem occurred while installing software.

> Send problem report to the developers?

> Package: onlyoffice-documentserver 5.6.1-6

What? I had _5.6.0-17_, right? So it seems to me that after the first install, _apt_ got a newer version information and couldn't completely remove the installed older version because of that? And I've got another error message:

> The problem cannot be reported:
> This is not an official Ubuntu package. Please remove any third party package and try again.

So what I think now is that DEB-file isn't configured _Ubuntu_-friendly.

Now, _apt-get_ way:
`peacecop@peacecop-HP-ProBook-4520s:~$ sudo apt-get install onlyoffice-documentserver
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  onlyoffice-documentserver
0 upgraded, 1 newly installed, 0 to remove and 5 not upgraded.
Need to get 261 MB of archives.
After this operation, 1 004 MB of additional disk space will be used.
Get:1 https://download.onlyoffice.com/repo/debian squeeze/main amd64 onlyoffice-documentserver amd64 5.6.1-6 [261 MB]
Fetched 261 MB in 3min 41s (1 181 kB/s)                                                                                                      
Preconfiguring packages ...
(Reading database ... 354363 files and directories currently installed.)
Preparing to unpack .../onlyoffice-documentserver_5.6.1-6_amd64.deb ...
Unpacking onlyoffice-documentserver (5.6.1-6) ...
Setting up onlyoffice-documentserver (5.6.1-6) ...
Generating AllFonts.js, please wait...Done
Generating presentation themes, please wait...Done
Congratulations, the ONLYOFFICE documentserver has been installed successfully!
Processing triggers for libc-bin (2.31-0ubuntu9) ...
`
There was no interaction with me with _apt-get_ neither. I couldn't choose which database engine to use. So you tell me that the options must be there because the user can use but the user actually can't choose between them and are forced to use the default _postgres_. Am I correct?

Another questions is: If the user could choose then why not at the start of the install not after it? For instance, I order a mean for sleeping. First, there will be a delivery with a bed and a hammock. Then I choose bed and the bed will be installed and hammock delivered back. Isn't it inefficient? I could choose between the bed and hammock in the ordering process, so that only the bed would be delivered.

**Comment by piiskop** [comment]
I figured out how the user can choose the database engine: _sudo dpkg-reconfigure onlyoffice-documentserver_. But this can be done after installing it wasting resources for one additional database engine. The user could choose the engine at the start of the installing process before anything will be downloaded. Don't you think so?

**Comment by piiskop** [comment]
I tested the interactive recofinguration by choosing _mariadb_, setting _localhost_ for the host, _onlyoffice_ for both the database and user name and password and _RabbitMQ_ password and left _guest_ as the user name for the latter. The result was:
`peacecop@peacecop-HP-ProBook-4520s:~$ sudo dpkg-reconfigure onlyoffice-documentserver 
[sudo] password for peacecop: 
Preparing for shutdown, it can take a lot of time, please wait...Done
Stopping documentserver services...
ds:docservice: stopped
ds:converter: stopped
ds:spellchecker: stopped
ds:metrics: stopped
ERRORERROR 1045 (28000): Access denied for user 'onlyoffice'@'localhost' (using password: YES)
 1045 (28000): Access denied for user 'onlyoffice'@'localhost' (using password: YES)
`
Especially interesting is here that this ended up not only by an error but an errorerror. I see that it complains about missing privileges for _mariadb_. Shall I set these privileges manually before reconfiguring? Why is it so that for _postgres_, everything will be done for me but for _mariadb_ not?

**Comment by piiskop** [comment]
Now, I tested whether the reconfiguring works after doing the dirty work manually before that:
`peacecop@peacecop-HP-ProBook-4520s:~$ sudo mysql -u root -p
Enter password: 
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 45
Server version: 10.3.22-MariaDB-1ubuntu1 Ubuntu 20.04

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| kalku20            |
| mysql              |
| performance_schema |
+--------------------+
4 rows in set (0.018 sec)

MariaDB [(none)]> create database oo character set utf8 collate utf8_general_ci;
Query OK, 1 row affected (0.003 sec)

MariaDB [(none)]> create user oo@localhost identified by 'onlyoffice';
Query OK, 0 rows affected (0.026 sec)

MariaDB [(none)]> grant all privileges on oo.* to oo@localhost;
Query OK, 0 rows affected (0.003 sec)

MariaDB [(none)]> flush privileges;
Query OK, 0 rows affected (0.008 sec)

MariaDB [(none)]> exit;
Bye
peacecop@peacecop-HP-ProBook-4520s:~$ sudo dpkg-reconfigure onlyoffice-documentserver 
Preparing for shutdown, it can take a lot of time, please wait...Done
Stopping documentserver services...
Generating AllFonts.js, please wait...Done
Generating presentation themes, please wait...Done
Congratulations, the ONLYOFFICE documentserver has been installed successfully!
Processing triggers for libc-bin (2.31-0ubuntu9) ...
`
So far so good. What about checking out in the web browser? The result is:

> Something went wrong during installation

**Comment by ShockwaveNN** [comment]
I agree with some of your points, but I don't like your tone.

So for me discussion is over, maybe some one else will be willing to help

I'll leave issue opened in that case

**Comment by markuman** [comment]
You can configure onlyoffice documentserver to work with a mysql (derivat (mariadb, percona)) database or with a postgres.  
The default configuration is postgres.  
So to make it easier, they pack both dependencies in their .deb package. Maybe it's easier to run integration tests adhoc against both. ...just a guess.

> Aren't they useless if the respective database server is not installed?

I guess (again), it depends on your setup. The db itself does not need be installed in the same host, but the clients must be.

By the way @piiskop do not use `utf8` on mysql/mariadb/percona. What you want is `utf8mb4`.

**Comment by piiskop** [comment]
Who are _they_ who pack these dependencies? Is there another, alien crew behind ONLYOFFICE whose members aren't part of these discussions here? Why isn't it documented?

What bad can happen if I use _utf8_? What's the quickest and safest way to convert the database to _utf8mb4_?

**Comment by Tachi107** [comment]
> Who are they who pack these dependencies? Is there another, alien crew behind ONLYOFFICE whose members aren't part of these discussions here?

I guess that not every member of the team knows everything about the DocumentServer, as OnlyOffice is a big project, so things like this can happen. It is also true that maybe the developers should be more open to discussions and listen to feedbacks from their users. 


> What bad can happen if I use _utf8_? What's the quickest and safest way to convert the database to _utf8mb4_?

In MySQL utf8 isn't really UTF8. It can cause some problems (expecially with Nextcloud), and you should switch to utf8mb4. You can learn more [here](https://docs.nextcloud.com/server/stable/admin_manual/configuration_database/mysql_4byte_support.html).

**Comment by jbhanks** [comment]
I am also a bit confused about how to set up with mysql. I installed the deb on Ubuntu 20.04  [as per these instructions:](https://helpcenter.onlyoffice.com/installation/docs-community-install-ubuntu.aspx). There were database connection errors during installation (because I set up the database with mysql not postgres). Since the expected config files had been installed, I changed the database type and encoding as described in this thread.

However, now I am not sure what to do. I don't have any process named "onlyoffice" running. I am guessing that it did not start correctly because it was configured by default to use postgres. But I see nothing in the documentation about how to start onlyoffice manually now that I have edited the configuration. There must be a command to start and stop it, right?

In looking for an answer I found this less-than-helpful thread: https://dev.onlyoffice.org/viewtopic.php?f=44&t=9982

**Comment by ShockwaveNN** [comment]
@jbhanks Hi, I think it's better to create new issue for your configuration, since your issue is not discussion about two databases, but seems something goes wrong during installation
